repository/population.go

```diff
@@ -0,0 +1,116 @@
+	mysql *sqlx.DB
+}
+
+type PopulationRepository interface {
+	QueryAllDistrict() ([]ResultQueryDistrictIdAndName, error)
+	QueryTotalPopulation(districtId string) (int64, error)
+	QueryPeopleCommitedTheVote(districtId string) (int64, error)
+	QueryPeopleRightToVote(districtId string) (int64, error)
+}
+
+func NewPopulationRepository(mysql *sqlx.DB) PopulationRepository {
+	return &populationRepository{
+		mysql: mysql,
+	}
+}
+
+type ResultQueryDistrictIdAndName struct {
+	LocationID string `db:"DistrictID"`
+	Location   string `db:"Name"`
+}
+
+func (p *populationRepository) queryLog(message string, functionName string) {
+	logrus.WithFields(logrus.Fields{
+		"Module":  "Repository",
+		"Funtion": functionName,
+	}).Info(message)
+}
+
+func (p *populationRepository) errorMessage(err error, functionName string) {
+	logrus.WithFields(logrus.Fields{
+		"Module":  "Repository",
+		"Funtion": functionName,
+	}).Error(err)
+}
+
+func (p *populationRepository) QueryAllDistrict() ([]ResultQueryDistrictIdAndName, error) {
+	res := []ResultQueryDistrictIdAndName{}
+	q := `SELECT Name, DistrictID
+	FROM District`
+	err := p.mysql.Select(&res, q)
+	if err != nil {
+		p.errorMessage(err, "QueryAllDistrict")
+		return res, err
+	}
+	p.queryLog("Select all district", "QueryAllDistrict")
+	return res, nil
+}
+func (p *populationRepository) QueryTotalPopulation(districtId string) (int64, error) {
+	var res int64
+	q := `
+	SELECT COUNT(*) as Total
+	FROM Population as p
+	WHERE DistrictId=?`
+	err := p.mysql.Get(&res, q, districtId)
+	if err != nil {
+		p.errorMessage(err, "QueryTotalPopulation")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total population districtId: %s", districtId), "QueryTotalPopulation")
+	return res, nil
+
+}
+
+func (p *populationRepository) QueryPeopleCommitedTheVote(districtId string) (int64, error) {
+	var res int64
+	q := `
+	SELECT COUNT(*) as Commits
+	FROM 
+	(
+		SELECT p.CitizenID, p.DistrictID
+		FROM Population as p 
+		JOIN ApplyVote as a 
+		on a.CitizenID = p.CitizenID
+	) as b
+	WHERE b.DistrictId = ?
+	`
+	err := p.mysql.Get(&res, q, districtId)
+	if err != nil {
+		p.errorMessage(err, "QueryPeopleCommitedTheVote")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total people commited the vote districtId: %s", districtId), "QueryPeopleCommitedTheVote")
+	return res, nil
+}
+
+func (p *populationRepository) QueryPeopleRightToVote(districtId string) (int64, error) {
+	var res int64
+	q := `
+	SELECT COUNT(*) as HaveRight
+	FROM (
+			SELECT * FROM Population AS p
+			WHERE p.Birthday < DATE_ADD(CURRENT_DATE(), INTERVAL -18 YEAR) 
+			AND p.CitizenID NOT IN (SELECT CitizenID FROM ApplyVote)
+			AND p.DistrictId=?
+			AND p.Nationality="Thai"
+		) AS c
+	`
+	err := p.mysql.Get(&res, q, districtId)
+	if err != nil {
+		p.errorMessage(err, "QueryPeopleRightToVote")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total people have right to vote districtId: %s", districtId), "QueryPeopleRightToVote")
```

Comment:
Will it better if we make the logging to start with a function name? ex. ("QueryPeopleRightToVote: Get total people have right to vote districtId: 1")

Reply:
<img width="803" alt="Screen Shot 2565-11-01 at 19 24 29" src="https://user-images.githubusercontent.com/69972884/199231678-81a5b84a-d6ff-40cd-9ba9-2ba8b99c7653.png">
It also has a function-name, and module name displayed on the back of the log message. Is this ok for you?

Reply:
Oh, this should work!